docs/source/customizations/native.mdx

```diff
@@ -183,107 +135,129 @@ impl Plugin for HelloWorld {
 
-Before implementing a layer yourself, always check if an existing layer implementation will fit your needs. Layer reuse is significantly easier than implementing layers from scratch.
+Sometimes you might need to pass custom information between services. For example:
 
-### Context
+* Authentication information obtained by the `RouterService` might be required by `SubgraphService`s.
+* Cache control headers from `SubgraphService`s might be aggregated and returned to the client by the `RouterService`.
 
-Sometimes there is a need to propagate information between services. For instance:
+Whenever the Apollo Router receives a request, it creates a corresponding `context` object and passes it along to each service. This object can store anything that is Serde-compatible (e.g., all simple types or a custom type).
 
-* Authentication information obtained at the `RouterService` is picked up by `SubgraphService(s)`.
-* Cache control headers at subgraph services are aggregated and returned to the client at the `RouterService`
+All of your plugin's hooks can interact with the `context` object using the following functions:
 
-When a request is received a context is created and propagated throughout the entire request. It can store anything that is Serde compatible. For instance, all simple types or a custom type.
+#### `insert`
 
-#### Insert
 ```rust
 context.insert("key1", 1)?;
-Serialization and deserialization happen automatically. You may sometimes need to specify the type in the case where the rust compiler cannot figure it out by itself.
 ```
 
-#### Get
+Adds a value to the `context` object. Serialization and deserialization happen automatically. You might sometimes need to specify the type in cases where the Rust compiler can't figure it out by itself.
+
+> If multiple threads might write a value to the same `context` key, use [`upsert`](#upsert) instead.
+
+#### `get`
+
 ```rust
 let value : u32 = context.get("key1")?;
-Upserts are guaranteed to happen serially. This is useful at the subgraph service where multiple requests are happening in parallel.
 ```
 
-#### Upsert
+Fetches a value from the `context` object.
+
+#### `upsert`
+
 ```rust
 context.upsert("key1", |v| v + 1, || 0)?;
 ```
-Use `upsert` if you need to merge several values. Rust is multi-threaded, and you will get unexpected results if multiple threads write to context at the same time. `upasert` prevents issues by guaranteeing that modifications happen serially.
 
-### Registering your plugin
-To enable the Apollo Router to discover your plugin it must be registered.
+Use `upsert` if you might need to resolve multiple simultaneous writes to a single `context` key (this is most likely for the `subgraph_service` hook, because it might be called by multiple threads in parallel). Rust is multi-threaded, and you will get unexpected results if multiple threads write to `context` at the same time. This function prevents issues by guaranteeing that modifications happen serially.
```

Comment:
Wasn't sure if `upsert` is necessary when multiple threads might modify the same _field_, or if it's even necessary when different threads are modifying _different_ fields of `context`